pow/src/test/java/tech/pegasys/artemis/pow/Eth1DataManagerTest.java

```diff
@@ -0,0 +1,226 @@
+import org.web3j.utils.Numeric;
+import tech.pegasys.artemis.pow.event.CacheEth1BlockEvent;
+import tech.pegasys.artemis.util.async.AsyncRunner;
+import tech.pegasys.artemis.util.async.AsyncRunnerTest;
+import tech.pegasys.artemis.util.async.SafeFuture;
+import tech.pegasys.artemis.util.config.Constants;
+
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class Eth1DataManagerTest {
+
+  private final EventBus eventBus = mock(EventBus.class);
+  private Eth1DataManager eth1DataManager;
+  private Web3j web3j;
+  private DepositContractListener depositContractListener;
+
+  private final AsyncRunner asyncRunner = new AsyncRunnerTest();
+
+  private final Bytes32 HEX_STRING = Bytes32.fromHexString("0xdeadbeef");
+  private final UnsignedLong LATEST_BLOCK_TIMESTAMP =
+      UnsignedLong.valueOf(Instant.now().getEpochSecond());
+  private final UnsignedLong INCONSEQUENTIAL_BLOCK_NUMBER = UnsignedLong.valueOf(100000);
+  private final Request mockLatestBlockRequest =
+      mockBlockRequest(INCONSEQUENTIAL_BLOCK_NUMBER, LATEST_BLOCK_TIMESTAMP);
+
+  @BeforeEach
+  void setUp() {
+    web3j = mock(Web3j.class);
+    depositContractListener = mock(DepositContractListener.class);
+    when(depositContractListener.getDepositCount(any()))
+        .thenReturn(SafeFuture.completedFuture(UnsignedLong.valueOf(1234)));
+    when(depositContractListener.getDepositRoot(any()))
+        .thenReturn(SafeFuture.completedFuture(HEX_STRING));
+  }
+
+  @Test
+  void cacheStartup_blockActuallyInMidRange() {
+    UnsignedLong midRangeTimestamp =
+        Eth1DataManager.getCacheRangeUpperBound().minus(UnsignedLong.ONE);
+
+    UnsignedLong firstUpwardNumber = INCONSEQUENTIAL_BLOCK_NUMBER.plus(UnsignedLong.ONE);
+
+    UnsignedLong secondUpwardNumber = firstUpwardNumber.plus(UnsignedLong.ONE);
+    UnsignedLong secondUpwardTimestamp = UnsignedLong.MAX_VALUE;
+
+    UnsignedLong firstDownwardNumber = INCONSEQUENTIAL_BLOCK_NUMBER.minus(UnsignedLong.ONE);
+
+    UnsignedLong secondDownwardNumber = firstDownwardNumber.minus(UnsignedLong.ONE);
+    UnsignedLong secondDownwardTimestamp = UnsignedLong.ONE;
+
+    Request midRangeBlockRequest =
+        mockBlockRequest(INCONSEQUENTIAL_BLOCK_NUMBER, midRangeTimestamp);
+    Request firstUpwardBlockRequest = mockBlockRequest(firstUpwardNumber, midRangeTimestamp);
+    Request secondUpwardBlockRequest = mockBlockRequest(secondUpwardNumber, secondUpwardTimestamp);
+    Request firstDownwardBlockRequest = mockBlockRequest(firstDownwardNumber, midRangeTimestamp);
+    Request secondDownwardBlockRequest =
+        mockBlockRequest(secondDownwardNumber, secondDownwardTimestamp);
+
+    when(web3j.ethGetBlockByNumber(any(), eq(false)))
+        .thenReturn(mockLatestBlockRequest)
+        .thenReturn(midRangeBlockRequest)
+        .thenReturn(firstUpwardBlockRequest)
+        .thenReturn(secondUpwardBlockRequest)
+        .thenReturn(firstDownwardBlockRequest)
+        .thenReturn(secondDownwardBlockRequest);
```

Comment:
I don't think this is giving us a very realistic set of results from web3j.  For example when we initially request the latest block, it returns`INCONSEQUENTIAL_BLOCK_NUMBER` but then `firstUpwardNumber` is one block higher than that which it can't be because `INCONSEQUENTIAL_BLOCK_NUMBER` is the chain head.  Really `INCONSEQUENTIAL_BLOCK_NUMBER` should be `LATEST_BLOCK_NUMBER` and everything should be before that.

And both `INCONSEQUENTIAL_BLOKC_NUMBER` and `firstUpwardBlockRequest` have the same timestamp which also can't happen.

I suspect we need to do the math to work out what the blockchain actually needs to look like realistically. I'd say start the test with something like:
```
final List<CacheEth1BlockEvent > blocks = Arrays.asList(
 mockBlock(10, <timestamp>),
 mockBlock(11, <timestamp>),
 .. for each block in the relevant range ..
 mockBlock(25, <timestamp>));
```

Then you can setup the web3j mock to respond to any request by block number for those blocks specifically instead of using `any()` and just need to add in an expectation for getting the `latest` block which would return the last block in the list. And I would actually use plain numbers not constants for both block number and timestamps in that list for maximum readability.  You'll have to inject a `Clock` as I mentioned earlier so that you can have a fixed time rather than having to calculate things around `Instant.now()`.

Doing it this way means we have a very clear, readable list of the blocks involved in the chain we're trying to explore - easy to see that they're valid and work out why we're expecting what we are from the test.

One very useful trick you'll need to make this work instead of `any()` - `DefaultBlockParameter` doesn't provide useful `equals` methods so we need a custom argument matcher or mockito will never match the params.

```
private DefaultBlockParameter blockParameterEq(final DefaultBlockParameter expected) {
    return argThat(arg -> arg != null && arg.getValue().equals(expected.getValue()));
  }
```

which would be used like:
`when(web3j.ethGetBlockByNumber(blockParameterEq(blockParam), eq(false))).thenReturn(request);`

Reply:
Done.